Mathlib/Analysis/Normed/Ring/Basic.lean

```diff
@@ -736,7 +730,44 @@ theorem norm_prod (s : Finset β) (f : β → α) : ‖∏ b ∈ s, f b‖ = ∏
 theorem nnnorm_prod (s : Finset β) (f : β → α) : ‖∏ b ∈ s, f b‖₊ = ∏ b ∈ s, ‖f b‖₊ :=
   map_prod nnnormHom.toMonoidHom f s
 
-end NormedCommRing
+end SeminormedCommRing
+
+section NormedAddCommGroup
+
+variable [NormedAddCommGroup α] [MulOneClass α] [NormMulClass α] [Nontrivial α]
+
+/-- Deduce `NormOneClass` from `NormMulClass` under a suitable nontriviality hypothesis. Not
+an instance, in order to avoid loops. -/
+lemma NormMulClass.toNormOneClass : NormOneClass α := by
+  constructor
+  obtain ⟨u, hu⟩ := exists_ne (0 : α)
+  simpa [mul_eq_left₀ (norm_ne_zero_iff.mpr hu)] using (norm_mul u 1).symm
+
+end NormedAddCommGroup
+
+section NormedRing
+
+variable [NormedRing α] [NormMulClass α]
+
+instance NormMulClass.isAbsoluteValue_norm : IsAbsoluteValue (norm : α → ℝ) where
+  abv_nonneg' := norm_nonneg
+  abv_eq_zero' := norm_eq_zero
+  abv_add' := norm_add_le
+  abv_mul' := norm_mul
+
+/-- A non-zero normed ring satisfying `NormMulClass` is a domain. Not an instance, to avoid
+loops. -/
+lemma NormMulClass.isDomain [Nontrivial α] : IsDomain α where
```

Comment:
I think if you demoted this to an `IsCancelMulZero` instance then the loop would go away (you could of course keep both, with the `IsDomain` as a lemma only)

Reply:
Better still, one can prove `NoZeroDivisors α`, which does not either require or imply `Nontrivial`, and then the prof of the non-instance `toIsDomain` is a one-liner.